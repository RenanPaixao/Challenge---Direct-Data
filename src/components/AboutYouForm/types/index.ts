export interface FieldConfig<t> {
  name: keyof t,
  type: string,
  label: string
}
