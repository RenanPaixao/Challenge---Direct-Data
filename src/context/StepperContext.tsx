import { UseStepsReturn } from '@chakra-ui/react'
import { createContext, PropsWithChildren } from 'react'
import { IStep } from '../components/TheStepper/TheStepper.tsx'

interface IProps extends UseStepsReturn {
  steps: IStep[]
}
export const StepperContext = createContext<IProps | null>(null)

StepperContext.displayName = 'StepperContext'

export const StepperProvider = ({ children, ...rest }: PropsWithChildren<IProps>) => {
  return <StepperContext.Provider value={{ ...rest }}>
    {children}
  </StepperContext.Provider>
}

