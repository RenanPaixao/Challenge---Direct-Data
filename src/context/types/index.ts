
export interface AboutYouInformation extends IdentityInformation {
  responsible: IdentityInformation | null
  weight: string
  height: string
  email: string
  phone: string
}

export interface Address {
  street: string,
  number: string,
  complement?: string,
  district: string,
  cep: string,
  city: string,
  state: string
}

export interface IdentityInformation {
  name: string ,
  lastName: string ,
  birthDate: string,
  cpf: string
}
