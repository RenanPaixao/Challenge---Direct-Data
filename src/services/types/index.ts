export interface StudentModel extends IdentityInformationModel {
  responsavel: IdentityInformationModel | null,
  peso: number,
  altura: number,
  email: string,
  telefoneContato: string,
  logradouro: string,
  numero: number,
  complemento: string,
  bairro: string,
  cep: string,
  cidade: string,
  uf: string
}

interface IdentityInformationModel {
  nome: string ,
  sobrenome: string ,
  dataNascimento: string,
  cpf: string
}
