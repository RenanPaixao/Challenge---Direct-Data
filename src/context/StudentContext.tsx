import { createContext, PropsWithChildren, useEffect, useState } from 'react'
import { AboutYouInformation } from './types'
import { persistOnSessionStorage, retrieveFromSessionStorage } from '../utils/persistence.ts'

export interface IProps {
  aboutYouInformation: AboutYouInformation | null
  setAboutYouInformation: (aboutYouInformation: AboutYouInformation) => void
}

export const StudentContext = createContext<IProps>({
  aboutYouInformation: null,
  setAboutYouInformation: () => {}
})

StudentContext.displayName = 'StudentContext'

export const StudentProvider = ({ children }: PropsWithChildren) => {
  const sessionStorageKey = 'aboutYouInformation'
  const [aboutYouInformation, setAboutYouInformation] = useState<AboutYouInformation | null>(() => {
    return retrieveFromSessionStorage(sessionStorageKey)
  })

  useEffect(() => {
    persistOnSessionStorage(sessionStorageKey, aboutYouInformation)
  }, [aboutYouInformation])

  return <StudentContext.Provider value={{
    aboutYouInformation,
    setAboutYouInformation
  }}>
    {children}
  </StudentContext.Provider>
}
