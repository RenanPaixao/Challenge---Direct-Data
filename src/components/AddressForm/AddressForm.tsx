import { Button, Center, SimpleGrid, SimpleGridProps } from '@chakra-ui/react'
import { renderInputsBaseOnConfigs } from '../../utils/forms.tsx'
import { FormFooterWrapper } from '../FormFooterWrapper/FormFooterWrapper.tsx'
import { FormikProvider, useFormik } from 'formik'
import { useContext } from 'react'
import { StepperContext } from '../../context/StepperContext.tsx'
import { FieldConfig } from '../AboutYouForm/types'
import * as Yup from 'yup'
import { FORM_MESSAGES } from '../../utils/constants.ts'
import { StudentContext } from '../../context/StudentContext.tsx'

const { REQUIRED, STATE_MIN_LENGTH } = FORM_MESSAGES

const schema = Yup.object({
  cep: Yup.string().required(REQUIRED),
  street: Yup.string().required(REQUIRED),
  district: Yup.string().required(REQUIRED),
  number: Yup.string().required(REQUIRED),
  city: Yup.string().required(REQUIRED),
  state: Yup.string().required(REQUIRED).min(2, STATE_MIN_LENGTH),
  complement: Yup.string()
})

type FormValues = Yup.InferType<typeof schema>

const fieldsConfig: FieldConfig[] = [
  {
    name: 'cep',
    type: 'text',
    label: 'CEP',
    mask: '00000-000'
  },
  {
    name: 'street',
    type: 'text',
    label: 'Logradouro'
  },
  {
    name: 'district',
    type: 'text',
    label: 'Bairro'
  },
  {
    name: 'number',
    type: 'text',
    label: 'Número',
    mask: Number
  },
  {
    name: 'city',
    type: 'text',
    label: 'Cidade'
  },
  {
    name: 'state',
    type: 'text',
    label: 'Estado',
    mask: 'aa'
  },
  {
    name: 'complement',
    type: 'text',
    label: 'Complemento'
  }
]
export const AddressForm = (props: SimpleGridProps) => {
  const { goToPrevious, goToNext } = useContext(StepperContext)!
  const { address, setAddress } = useContext(StudentContext)

  const formik = useFormik<FormValues>({
    initialValues: {
      cep: '',
      city: '',
      complement: '',
      district: '',
      street: '',
      state: '',
      number: '',
      ...address
    },
    onSubmit: values => {
      setAddress(values)
      goToNext()
    },
    validationSchema: schema
  })

  return <FormikProvider value={formik}>
    <Center as={'form'} onSubmit={formik.handleSubmit} flexDirection={'column'}>
      <SimpleGrid {...props} columns={[1, 2]} spacing={8}>
        {renderInputsBaseOnConfigs(fieldsConfig)}
      </SimpleGrid>
      <FormFooterWrapper pt={10}>
        <Button onClick={goToPrevious}>
          Voltar
        </Button>
        <Button type={'submit'} isLoading={formik.isSubmitting}>
        Avançar
        </Button>
      </FormFooterWrapper>
    </Center>
  </FormikProvider>
}
