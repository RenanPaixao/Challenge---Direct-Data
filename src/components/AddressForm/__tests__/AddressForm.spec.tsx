
import { fillForm, renderWithStudentContext } from '../../../../tests/test-utils.tsx'
import { Address } from '../../../context/types'
import { AddressForm } from '../AddressForm.tsx'

interface FormField {
  name: string
  value: string
  role: string
}

type FormValues = Record<keyof Omit<Address, 'responsible'>, FormField>

const formValues: FormValues = {
  cep: {
    name: 'cep',
    value: '12345678',
    role: 'textbox'
  },
  number: {
    name: 'number',
    value: '15',
    role: 'textbox'
  },
  city: {
    name: 'city',
    value: 'test city',
    role: 'textbox'
  },
  state: {
    name: 'state',
    value: 'al',
    role: 'textbox'
  },
  complement: {
    name: 'complement',
    value: 'test complement',
    role: 'textbox'
  },
  street: {
    name: 'street',
    value: 'test street',
    role: 'textbox'
  },
  district: {
    name: 'district',
    value: 'test district',
    role: 'textbox'
  }
}

describe('AboutYouForm', () => {
  it('should render', async () => {
    const { container } = renderWithStudentContext(<AddressForm/>)

    expect(container).toMatchSnapshot()
  })

  it('should have masks working', async () => {
    const {  getByRole } = renderWithStudentContext(<AddressForm />)
    const { cep, number, state } = formValues

    await fillForm<FormValues>({
      cep: {
        ...cep,
        value: 'aaa' + cep.value + '2aaa2'
      },
      number: {
        ...number,
        value: 'aaa' + number.value + 'aaa'
      },
      state: {
        ...state,
        value: state.value + 'aaa'
      }
    })

    const cepField = getByRole(cep.role, { name: cep.name })
    const numberField = getByRole(number.role, { name: number.name })
    const stateField = getByRole(state.role, { name: state.name })

    expect(cepField).toHaveValue(cep.value.substring(0,5) + '-' + cep.value.substring(5))
    expect(numberField).toHaveValue(number.value)
    expect(stateField).toHaveValue(state.value)
  })
})
