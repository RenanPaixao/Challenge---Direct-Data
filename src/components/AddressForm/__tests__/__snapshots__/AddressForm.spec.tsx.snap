// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddressForm > should render 1`] = `
<div>
  <form
    class="css-b93e9j"
  >
    <div
      class="css-1gkent6"
    >
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="cep"
          id="field-:r0:-label"
        >
          CEP
        </label>
        <input
          aria-describedby="field-:r0:-helptext"
          aria-label="cep"
          class="chakra-input css-1cjy4zv"
          id="field-:r0:"
          name="cep"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r0:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="street"
          id="field-:r1:-label"
        >
          Logradouro
        </label>
        <input
          aria-describedby="field-:r1:-helptext"
          aria-label="street"
          class="chakra-input css-1cjy4zv"
          id="field-:r1:"
          name="street"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r1:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="district"
          id="field-:r2:-label"
        >
          Bairro
        </label>
        <input
          aria-describedby="field-:r2:-helptext"
          aria-label="district"
          class="chakra-input css-1cjy4zv"
          id="field-:r2:"
          name="district"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r2:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="number"
          id="field-:r3:-label"
        >
          Número
        </label>
        <input
          aria-describedby="field-:r3:-helptext"
          aria-label="number"
          class="chakra-input css-1cjy4zv"
          id="field-:r3:"
          name="number"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r3:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="city"
          id="field-:r4:-label"
        >
          Cidade
        </label>
        <input
          aria-describedby="field-:r4:-helptext"
          aria-label="city"
          class="chakra-input css-1cjy4zv"
          id="field-:r4:"
          name="city"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r4:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="state"
          id="field-:r5:-label"
        >
          Estado
        </label>
        <input
          aria-describedby="field-:r5:-helptext"
          aria-label="state"
          class="chakra-input css-1cjy4zv"
          id="field-:r5:"
          name="state"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r5:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="complement"
          id="field-:r6:-label"
        >
          Complemento
        </label>
        <input
          aria-describedby="field-:r6:-helptext"
          aria-label="complement"
          class="chakra-input css-1cjy4zv"
          id="field-:r6:"
          name="complement"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r6:-helptext"
        />
      </div>
    </div>
    <div
      class="css-1udc1ya"
    >
      <button
        class="chakra-button css-ez23ye"
        type="button"
      >
        Voltar
      </button>
      <button
        class="chakra-button css-ez23ye"
        type="submit"
      >
        Avançar
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
